docs/man-openmpi/man1/mpirun.1.rst
.. _man1-mpirun:
.. _man1-mpiexec:


mpirun / mpiexec
================

.. include_body

mpirun, mpiexec |mdash| Execute serial and parallel jobs in Open MPI.

.. note:: ``mpirun`` and ``mpiexec`` are synonyms for each other.
          Indeed, they are symbolic links to the same executable.
          Using either of the names will produce the exact same
          behavior.

SYNOPSIS
--------

Single Process Multiple Data (SPMD) Model:

.. code::

   mpirun [ options ] <program> [ <args> ]

Multiple Instruction Multiple Data (MIMD) Model:

.. code::

   mpirun [ global_options ]
          [ local_options1 ] <program1> [ <args1> ] :
          [ local_options2 ] <program2> [ <args2> ] :
          ... :
          [ local_optionsN ] <programN> [ <argsN> ]

Note that in both models, invoking ``mpirun`` via an absolute path
name is equivalent to specifying the ``--prefix`` option with a
``<dir>`` value equivalent to the directory where ``mpirun`` resides,
minus its last subdirectory.  For example:

.. code:: sh

   shell$ /usr/local/bin/mpirun ...

is equivalent to

.. code:: sh

   shell$ mpirun --prefix /usr/local

QUICK SUMMARY
-------------

If you are simply looking for how to run an MPI application, you
probably want to use a command line of the following form:

.. code:: sh

   shell$ mpirun [ -n X ] [ --hostfile <filename> ]  <program>

This will run ``X`` copies of ``<program>`` in your current run-time
environment (if running under a supported resource manager, Open MPI's
``mpirun`` will usually automatically use the corresponding resource
manager process starter, as opposed to ``ssh`` (for example), which
require the use of a hostfile, or will default to running all ``X``
copies on the localhost), scheduling (by default) in a round-robin
fashion by CPU slot.  See the rest of this documentation for more
details.

Please note that ``mpirun`` automatically binds processes to hardware
resources. Three binding patterns are used in the absence of any
further directives (See :ref:`map/rank/bind defaults
<man1-mpirun-map-rank-bind-defaults>` for more details):

* **Bind to core**:     when the number of processes is <= 2
* **Bind to package**:  when the number of processes is > 2
* **Bind to none**:     when oversubscribed

If your application uses threads, then you probably want to ensure
that you are either not bound at all (by specifying ``--bind-to none``),
or bound to multiple cores using an appropriate binding level or
specific number of processing elements per application process.

OPEN MPI'S USE OF PRRTE
-----------------------

Open MPI uses the PMIx Reference Runtime Environment (PRRTE) as the
main engine for launching, monitoring, and terminating MPI processes.

Much of the documentation below is directly imported from PRRTE.  As
such, it frequently refers to PRRTE concepts and command line options.
Except where noted, these concepts and command line argument are all
applicable to Open MPI as well.  Open MPI extends the available PRRTE
command line options, and also slightly modifies the PRRTE's default
behaviors in a few cases.  These will be specifically described in the
documentation below.

COMMAND LINE OPTIONS
--------------------

The core of Open MPI's ``mpirun`` processing is performed via the
`PRRTE <https://docs.prrte.org/>`_.  Specifically: ``mpirun`` is
effectively a wrapper around ``prterun``, but ``mpirun``'s CLI options
are slightly different than PRRTE's CLI commands.

.. include:: /schizo-ompi-rst-content/schizo-ompi-cli.rstxt

OPTIONS (OLD / HARD-CODED CONTENT |mdash| TO BE AUDITED
-------------------------------------------------------

.. admonition:: This is old content
   :class: error

   This is the old section of manually hard-coded content.  It should
   probably be read / audited and see what we want to keep and what we
   want to discard.

   Feel free to refer to https://docs.prrte.org/ rather than
   replicating content here (e.g., for the definition of a slot and
   other things).

mpirun will send the name of the directory where it was invoked on the
local node to each of the remote nodes, and attempt to change to that
directory.  See the :ref:`man1-mpirun-current-working-directory`
section below for further details.

* ``<program>``: The program executable. This is identified as the
  first non-recognized argument to mpirun.

* ``<args>``: Pass these run-time arguments to every new process.
  These must always be the last arguments to mpirun. If an app context
  file is used, ``<args>`` will be ignored.

* ``-h``, ``--help``: Display help for this command

* ``-q``, ``--quiet``: Suppress informative messages from orterun
  during application execution.

* ``-v``, ``--verbose``:` Be verbose

* ``-V``, ``--version``: Print version number.  If no other arguments
  are given, this will also cause orterun to exit.

* ``-N <num>``: Launch num processes per node on all allocated nodes
  (synonym for ``--npernode``).

* ``--display-map``: Display a table showing the mapped location of
  each process prior to launch.

* ``--display-allocation``: Display the detected resource allocation.

* ``--output-proctable``: Output the debugger proctable after launch.

* ``--dvm``: Create a persistent distributed virtual machine (DVM).

* ``--max-vm-size <size>``: Number of daemons to start.

Use one of the following options to specify which hosts (nodes) of the
cluster to run on. Note that as of the start of the v1.8 release,
mpirun will launch a daemon onto each host in the allocation (as
modified by the following options) at the very beginning of execution,
regardless of whether or not application processes will eventually be
mapped to execute there. This is done to allow collection of hardware
topology information from the remote nodes, thus allowing us to map
processes against known topology. However, it is a change from the
behavior in prior releases where daemons were only launched after
mapping was complete, and thus only occurred on nodes where
application processes would actually be executing.

* ``-H``, ``--host <host1,host2,...,hostN>``: list of hosts on which to
  invoke processes.

* ``--hostfile <hostfile>``: Provide a hostfile to use.

* ``--default-hostfile <hostfile>``: Provide a default hostfile.

* ``--machinefile <machinefile>``: Synonym for ``--hostfile``.

* ``--cpu-set <list>``: Restrict launched processes to the specified
  logical CPUs on each node (comma-separated list). Note that the
  binding options will still apply within the specified envelope
  |mdash| e.g., you can elect to bind each process to only one CPU
  within the specified CPU set.

The following options specify the number of processes to launch. Note
that none of the options imply a particular binding policy |mdash| e.g.,
requesting N processes for each package does not imply that the
processes will be bound to the package.

* ``-n``, ``--n``, ``-c``, ``-np <#>``: Run this many copies of the
  program on the given nodes.  This option indicates that the
  specified file is an executable program and not an application
  context. If no value is provided for the number of copies to execute
  (i.e., neither the ``-n`` nor its synonyms are provided on the
  command line), Open MPI will automatically execute a copy of the
  program on each process slot (see PRRTE's `definition of "slot"
  <https://docs.prrte.org/en/latest/placement/overview.html#definition-of-slot>`_
  for description of a "process slot"). This feature, however, can
  only be used in the SPMD model and will return an error (without
  beginning execution of the application) otherwise.

  .. note:: The ``-n`` option is the preferred option to be used to specify the
            number of copies of the program to be executed, but the alternate
            options are also accepted.


* ``--map-by ppr:N:<object>``: Launch N times the number of objects of
  the specified type on each node.

* ``--npersocket <#persocket>``: On each node, launch this many
  processes times the number of processor sockets on the node.
  The -npersocket option also turns on the ``--bind-to-socket``
  option.  (deprecated in favor of ``--map-by ppr:n:package``)

* ``--npernode <#pernode>``: On each node, launch this many processes.
  (deprecated in favor of ``--map-by ppr:n:node``).

* ``--pernode``: On each node, launch one process |mdash| equivalent to
  ``--npernode 1``.  (deprecated in favor of ``--map-by ppr:1:node``)

To map processes:

* ``--map-by <object>``: Map to the specified object, defaults to
  ``package``. Supported options include ``slot``, ``hwthread``, ``core``,
  ``L1cache``, ``L2cache``, ``L3cache``, ``package``, ``numa``,
  ``node``, ``seq``, ``rankfile``, ``pe-list=#``, and ``ppr``.
  Any object can include modifiers by adding a ``:`` and any combination
  of the following:

    * ``pe=n``: bind ``n`` processing elements to each proc
    * ``span``: load balance the processes across the allocation
    * ``oversubscribe``: allow more processes on a node than processing elements
    * ``nooversubscribe``: do *not* allow more processes on a node than processing elements (default)
    * ``nolocal``: do not place processes on the same host as the ``mpirun`` process
    * ``hwtcpus``: use hardware threads as CPU slots for mapping
    * ``corecpus``: use processor cores as CPU slots for mapping (default)
    * ``file=filename``: used with ``rankfile``; use ``filename`` to specify the file to use
    * ``ordered``: used with ``pe-list`` to bind each process to one of the specified processing elements

  .. note:: ``socket`` is also accepted as an alias for ``package``.

* ``--bycore``: Map processes by core (deprecated in favor of
  ``--map-by core``).

* ``--byslot``: Map and rank processes round-robin by slot (deprecated
  in favor of ``--map-by slot``).

* ``--nolocal``: Do not run any copies of the launched application on
  the same node as orterun is running.  This option will override
  listing the localhost with ``--host`` or any other host-specifying
  mechanism. Alias for ``--map-by :nolocal``.

* ``--nooversubscribe``: Do not oversubscribe any nodes; error
  (without starting any processes) if the requested number of
  processes would cause oversubscription.  This option implicitly sets
  "max_slots" equal to the "slots" value for each node. (Enabled by
  default). Alias for ``--map-by :nooversubscribe``.

* ``--oversubscribe``: Nodes are allowed to be oversubscribed, even on
  a managed system, and overloading of processing elements.
  Alias for ``--map-by :oversubscribe``.

* ``--bynode``: Launch processes one per node, cycling by node in a
  round-robin fashion.  This spreads processes evenly among nodes and
  assigns ``MPI_COMM_WORLD`` ranks in a round-robin, "by node" manner.
  (deprecated in favor of ``--map-by node``)

* ``--cpu-list <cpus>``: Comma-delimited list of processor IDs to
  which to bind processes [default=``NULL``].  Processor IDs are
  interpreted as hwloc logical core IDs.

  .. note:: You can run Run the hwloc ``lstopo(1)`` command to see a
            list of available cores and their logical IDs.

To order processes' ranks in ``MPI_COMM_WORLD``:

* ``--rank-by <mode>``: Rank in round-robin fashion according to the
  specified mode, defaults to slot. Supported options include
  ``slot``, ``node``, ``fill``, and ``span``.

For process binding:

* ``--bind-to <object>``: Bind processes to the specified object,
  defaults to ``core``.  Supported options include ``slot``,
  ``hwthread``, ``core``, ``l1cache``, ``l2cache``, ``l3cache``,
  ``package``, ``numa``, and ``none``.

* ``--cpus-per-proc <#perproc>``: Bind each process to the specified
  number of cpus.  (deprecated in favor of ``--map-by <obj>:PE=n``)

* ``--cpus-per-rank <#perrank>``: Alias for ``--cpus-per-proc``.
  (deprecated in favor of ``--map-by <obj>:PE=n``)

* ``--bind-to-core`` Bind processes to cores (deprecated in favor of
  ``--bind-to core``)

* ``--bind-to-socket``: Bind processes to processor sockets
  (deprecated in favor of ``--bind-to package``)

* ``--report-bindings``: Report any bindings for launched processes.

For rankfiles:

* ``--rankfile <rankfile>``: Provide a rankfile file.
  (deprecated in favor of ``--map-by rankfile:file=FILE``)

To manage standard I/O:

* ``--output-filename <filename>``: Redirect the stdout, stderr, and
  stddiag of all processes to a process-unique version of the
  specified filename. Any directories in the filename will
  automatically be created.  Each output file will consist of
  ``filename.id``, where the ``id`` will be the processes' rank in
  ``MPI_COMM_WORLD``, left-filled with zero's for correct ordering in
  listings. A relative path value will be converted to an absolute
  path based on the cwd where mpirun is executed. Note that this will
  not work on environments where the file system on compute nodes
  differs from that where :ref:`mpirun(1) <man1-mpirun>` is
  executed.

* ``--stdin <rank>``: The ``MPI_COMM_WORLD`` rank of the process that is
  to receive stdin.  The default is to forward stdin to ``MPI_COMM_WORLD``
  rank 0, but this option can be used to forward stdin to any
  process. It is also acceptable to specify none, indicating that no
  processes are to receive stdin.

* ``--merge-stderr-to-stdout``: Merge stderr to stdout for each
  process.

* ``--tag-output``: Tag each line of output to stdout, stderr, and
  stddiag with ``[jobid, MCW_rank]<stdxxx>`` indicating the process
  jobid and ``MPI_COMM_WORLD`` rank of the process that generated the
  output, and the channel which generated it.

* ``--timestamp-output``: Timestamp each line of output to stdout,
  stderr, and stddiag.

* ``--xml``: Provide all output to stdout, stderr, and stddiag in an
  XML format.

* ``--xml-file <filename>`` Provide all output in XML format to the
  specified file.

* ``--xterm <ranks>``: Display the output from the processes
  identified by their ``MPI_COMM_WORLD`` ranks in separate xterm
  windows. The ranks are specified as a comma-separated list of
  ranges, with a ``-1`` indicating all. A separate window will be created
  for each specified process.

  .. note:: xterm will normally terminate the window upon termination
            of the process running within it. However, by adding a
            ``!`` to the end of the list of specified ranks, the
            proper options will be provided to ensure that xterm keeps
            the window open after the process terminates, thus
            allowing you to see the process' output.  Each xterm
            window will subsequently need to be manually closed.
            Note: In some environments, xterm may require that the
            executable be in the user's path, or be specified in
            absolute or relative terms. Thus, it may be necessary to
            specify a local executable as ``./my_mpi_app`` instead of just
            ``my_mpi_app``. If xterm fails to find the executable, ``mpirun``
            will hang, but still respond correctly to a ctrl-C.  If
            this happens, please check that the executable is being
            specified correctly and try again.

To manage files and runtime environment:

* ``--path <path>``: ``<path>`` that will be used when attempting to
  locate the requested executables.  This is used prior to using the
  local ``PATH`` environment variable setting.

* ``--prefix <dir>``: Prefix directory that will be used to set the
  ``PATH`` and ``LD_LIBRARY_PATH`` on the remote node before invoking
  Open MPI or the target process.  See the :ref:`Remote Execution
  <man1-mpirun-remote-execution>` section, below.

* ``--noprefix``: Disable the automatic ``--prefix`` behavior

* ``--preload-binary``: Copy the specified executable(s) to remote
  machines prior to starting remote processes. The executables will be
  copied to the Open MPI session directory and will be deleted upon
  completion of the job.

* ``--preload-files <files>``: Preload the comma-separated list of
  files to the current working directory of the remote machines where
  processes will be launched prior to starting those processes.

* ``--set-cwd-to-session-dir``: Set the working directory of the
  started processes to their session directory.

* ``--wd <dir>``: Synonym for ``-wdir``.

* ``--wdir <dir>``: Change to the directory ``<dir>`` before the
  user's program executes.  See the :ref:`Current Working Directory
  <man1-mpirun-current-working-directory>` section for notes on
  relative paths.  Note: If the ``--wdir`` option appears both on the
  command line and in an application context, the context will take
  precedence over the command line. Thus, if the path to the desired
  wdir is different on the backend nodes, then it must be specified as
  an absolute path that is correct for the backend node.

* ``-x <env>``: Export the specified environment variables to the
  remote nodes before executing the program.  Only one environment
  variable can be specified per ``-x`` option.  Existing environment
  variables can be specified or new variable names specified with
  corresponding values.  For example:

  .. code:: sh

     shell$ mpirun -x DISPLAY -x OFILE=/tmp/out ...

  The parser for the ``-x`` option is not very sophisticated; it does
  not even understand quoted values.  Users are advised to set
  variables in the environment, and then use ``-x`` to export (not
  define) them.

Setting MCA parameters:

* ``--gmca <key> <value>``: Pass global MCA parameters that are
  applicable to all contexts.  ``<key>`` is the parameter name;
  ``<value>`` is the parameter value.

* ``--mca <key> <value>``: Send arguments to various MCA modules.  See
  the :ref:`Setting MCA Parameters
  <man1-mpirun-setting-mca-parameters>` section for more details.

  .. note:: Open MPI will attempt to discern PMIx and PRRTE MCA
            parameters passed via ``--mca`` and handle them
            appropriately, but it may not always guess correctly.  It
            is best to use ``--pmixmca`` and ``--prtemca`` when
            passing MCA parameters to PMIx and PRRTE, respectively.

* ``--pmixmca <key> <value>``: Send arguments to MCA modules in the
  PMIx subsystem.  See the :ref:`Setting MCA Parameters
  <man1-mpirun-setting-mca-parameters>` section for more details.

* ``--prtemca <key> <value>``: Send arguments to MCA modules in the
  PMIx Reference Runtime Environment (PRRTE) subsystem.  See the
  :ref:`Setting MCA Parameters <man1-mpirun-setting-mca-parameters>`
  section for more details.

* ``--tune <tune_file>``: Specify a tune file to set arguments for
  various MCA modules and environment variables.  See the :ref:`
  Setting MCA parameters and environment variables from file
  <man1-mpirun-setting-mca-params-from-file>`. ``--am <arg>`` is an alias for ``--tune <arg>``.

For debugging:

* ``--get-stack-traces``: When paired with the ``--timeout`` option,
  ``mpirun`` will obtain and print out stack traces from all launched
  processes that are still alive when the timeout expires.  Note that
  obtaining stack traces can take a little time and produce a lot of
  output, especially for large process-count jobs.

* ``--timeout <seconds>``: The maximum number of seconds that
  ``mpirun`` will run.  After this many seconds, ``mpirun`` will abort
  the launched job and exit with a non-zero exit status.  Using
  ``--timeout`` can be also useful when combined with the
  ``--get-stack-traces`` option.

There are also other options:

* ``--allow-run-as-root``: Allow ``mpirun`` to run when executed by
  the root user (``mpirun`` defaults to aborting when launched as the
  root user).  Be sure to see the :ref:`Running as root
  <man1-mpirun-running-as-root>` section for more detail.

* ``--app <appfile>``: Provide an appfile, ignoring all other command
  line options.

* ``--continuous``: Job is to run until explicitly terminated.

* ``--disable-recovery``: Disable recovery (resets all recovery
  options to off).

* ``--do-not-launch``: Perform all necessary operations to prepare to
  launch the application, but do not actually launch it.

* ``--enable-recovery``: Enable recovery from process failure (default:
  disabled)

* ``--leave-session-attached``: Do not detach back-end daemons used by
  this application. This allows error messages from the daemons as
  well as the underlying environment (e.g., when failing to launch a
  daemon) to be output.

* ``--max-restarts <num>``: Max number of times to restart a failed
  process.

* ``--personality <list>``: Comma-separated list of programming model,
  languages, and containers being used (default=``ompi``).

* ``--ppr <list>``: Comma-separated list of number of processes on a
  given resource type (default: none). Alias for ``--map-by ppr:N:OBJ``.

* ``--report-child-jobs-separately``: Return the exit status of the
  primary job only.

* ``--report-events <URI>``: Report events to a tool listening at the
  specified URI.

* ``--report-pid <channel>``: Print out ``mpirun``'s PID during
  startup. The channel must be either a ``-`` to indicate that the PID
  is to be output to stdout, a ``+`` to indicate that the PID is to be
  output to stderr, or a filename to which the PID is to be written.

* ``--report-uri <channel>``: Print out ``mpirun``'s URI during
  startup. The channel must be either a ``-`` to indicate that the URI
  is to be output to stdout, a ``+`` to indicate that the URI is to be
  output to stderr, or a filename to which the URI is to be written.

* ``--show-progress``: Output a brief periodic report on launch
  progress.

* ``--terminate``: Terminate the DVM.

* ``--use-hwthread-cpus``: Use hardware threads as independent CPUs.

  Note that if a number of slots is not provided to Open MPI (e.g.,
  via the ``slots`` keyword in a hostfile or from a resource manager
  such as Slurm), the use of this option changes the default
  calculation of number of slots on a node.  See the PRRTE's
  `definition of "slot"
  <https://docs.prrte.org/en/latest/placement/overview.html#definition-of-slot>`_
  for more details.

  Also note that the use of this option changes the Open MPI's
  definition of a "processor element" from a processor core to a
  hardware thread.  See
  PRRTE's `definition of a "processor element"
  <https://docs.prrte.org/en/latest/placement/overview.html#definition-of-pe>`_
  for more details.

The following options are useful for developers; they are not
generally useful to most Open MPI users:

* ``--debug-daemons``: Enable debugging of the run-time daemons used
  by this application.

* ``--debug-daemons-file``: Enable debugging of the run-time daemons
  used by this application, storing output in files.

* ``--display-devel-map``: Display a more detailed table showing the
  mapped location of each process prior to launch.

* ``--display-topo``: Display the topology as part of the process map
  just before launch.

* ``--launch-agent``: Name of the executable that is to be used to
  start processes on the remote nodes. The default is ``prted``. This
  option can be used to test new daemon concepts, or to pass options
  back to the daemons without having mpirun itself see them. For
  example, specifying a launch agent of ``prted --prtemca odls_base_verbose
  5`` allows the developer to ask the ``prted`` for debugging output
  without clutter from ``mpirun`` itself.

* ``--report-state-on-timeout``: When paired with the ``--timeout``
  command line option, report the run-time subsystem state of each
  process when the timeout expires.

There may be other options listed with ``mpirun --help``.

Environment Variables
^^^^^^^^^^^^^^^^^^^^^

.. admonition:: This is old, hard-coded content
   :class: error


NEWS
Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
                        University Research and Technology
                        Corporation.  All rights reserved.
Copyright (c) 2004-2006 The University of Tennessee and The University
                        of Tennessee Research Foundation.  All rights
                        reserved.
Copyright (c) 2004-2008 High Performance Computing Center Stuttgart,
                        University of Stuttgart.  All rights reserved.
Copyright (c) 2004-2006 The Regents of the University of California.
                        All rights reserved.
Copyright (c) 2006-2023 Cisco Systems, Inc.  All rights reserved.
Copyright (c) 2006      Voltaire, Inc. All rights reserved.
Copyright (c) 2006      Sun Microsystems, Inc.  All rights reserved.
                        Use is subject to license terms.
Copyright (c) 2006-2018 Los Alamos National Security, LLC.  All rights
                        reserved.
Copyright (c) 2010-2020 IBM Corporation.  All rights reserved.
Copyright (c) 2012      Oak Ridge National Labs.  All rights reserved.
Copyright (c) 2012      Sandia National Laboratories.  All rights reserved.
Copyright (c) 2012      University of Houston. All rights reserved.
Copyright (c) 2013      NVIDIA Corporation.  All rights reserved.
Copyright (c) 2013-2018 Intel, Inc. All rights reserved.
Copyright (c) 2018-2024 Amazon.com, Inc. or its affiliates.  All Rights
                        reserved.
Copyright (c) 2019-2021 Triad National Security, LLC. All rights
                        reserved.
Copyright (c) 2021      Google, LLC. All rights reserved.
Copyright (c) 2023      Jeffrey M. Squyres.  All rights reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of Open MPI since
version 1.0.

As more fully described in the "Software Version Number" section in
the README file, Open MPI typically releases two separate version
series simultaneously.  Since these series have different goals and
are semi-independent of each other, a single NEWS-worthy item may be
introduced into different series at different times.  For example,
feature F was introduced in the vA.B series at version vA.B.C, and was
later introduced into the vX.Y series at vX.Y.Z.

The first time feature F is released, the item will be listed in the
vA.B.C section, denoted as:

   (** also to appear: X.Y.Z) -- indicating that this item is also
                                 likely to be included in future release
                                 version vX.Y.Z.

When vX.Y.Z is later released, the same NEWS-worthy item will also be
included in the vX.Y.Z section and be denoted as:

   (** also appeared: A.B.C)  -- indicating that this item was previously
                                 included in release version vA.B.C.

4.1.8 -- January, 2024
----------------------

** NOTE: Open MPI 4.1.7 incorrectly changed the shared library
   versioning information for the OpenSHMEM interface, potentially
   causing link errors when updating from prior releases of Open MPI
   to 4.1.7.  This release resets the shared library version number
   history to allow updates from Open MPI 4.1.6 and prior.  Customers
   who updated to 4.1.7 and relinked their OpenSHMEM applications will
   have to relink their application again.

- Allocate and exchange OpenSHMEM base segment addresses earlier in
  startup.
- Fixed file_seek calculation when using SEEK_END and io/ompio.
- Protect against using CUDA without VMM support.
- Change several variables in coll/adapt, coll/basic, coll/han,
  coll/hcoll, and coll/ucc from READONLY to ALL scope so that they can
  be set via the MPI_T interface.
- Add version identifier to rules file in coll/tuned and extend the
  dynamic rules file to support the alltoall_algorithm_max_requests
  tuning parameter.


4.1.7 -- October, 2024
----------------------

- Improve CUDA memory pool and context handling.
- Fix detection of host vs. CUDA memory when cumemcreate is used with
  certain arguments, leading to a crash.
- Fix int overflow bug in UCC collectives routine, which would impact
  large message collectives.
- Fix SHMEM bounds checking warnings
- Fix memory leak in the fcoll/vulcan component
- Fix memory leak when using non-contiguious datatyles with pml/cm
  (used when using Libfabric or Portals transports)
- Change default locking protocol for ufs IO component to fix data
  sieving use case.
- Fix bug in Fortran / C interface array calculation.
- Fix potential initialization in UCX OpenSHMEM support
- Add OpenSHMEM support for symmetric remote key handling to improve
  scalability of OpenSHMEM interface.
- Add MCS Locks as default implementation for OpenSHMEM locks.
- Fix bug to properly handle ORTE Prefix of "/".
- Fix bug in I/O interface handling of soft links on Lustre
  filesystems.
- Fix memory leak in UCX one-sided implementation.
- Fix bug in rcache components that could lead to assert on macOS
  systems.
- Fix race condition in smcuda Shared Memory transport (used on
  platforms with CUDA support).
- Fix mishandling of a NULL completion context in the OFI transport.
- Tarballs are now built with the tar-pax format.
- Set MPI_THREAD_SERIALIZED instead of MPI_THREAD_SINGLE as supported
  threading model when using UCX that supports
  UCS_THREAD_MODE_SERIALIZED.
- Improve macOS configure test to only add `-ld_classic` to the
  Fortran bindings link line if using a version of Xcode that ships a
  version of the "new linker" that does not support `-commons
  use_dylibs`.  Xcode 16 betas ship a new linker that supports
  `-commons use_dylibs`.
- Fix issue that would prevent using Libfabric 2.x releases.


4.1.6 -- September, 2023
------------------------

- Fix configure issue with XCode 15.

contrib/completion/mpirun.sh
#!/bin/bash # -*- Mode: shell-script; indent-tabs-mode:nil -*-
# SPDX-License-Identifier: BSD-3-Clause-Open-MPI
# (c) 2008-2013 Nathan Hjelm <hjelmn@cs.unm.edu>
# mpirun completion v1.1
#
# Bash completion script for Open MPI's mpirun
#
# Installation:
#   If bash completion is not already installed, follow the instructions at:
#      http://anonscm.debian.org/gitweb/?p=bash-completion/bash-completion.git;a=blob_plain;f=README
#
# Then install mpirun completion by dropping this file into either the system bash_completion.d
# (usually found in /etc) or somewhere in your home directory (~/.bash_completion.d is a good place).
# If installing the script in a location other than the system-wide location, you will probably need
# to add the following to either ~/.bash_profile or ~/.bash_completion:
#
#        [ -r /path/to/mpirun.sh ] && . /path/to/mpirun.sh

# find available MCA variables
_get_mca_variable_names() {
    ompi_info -a --parsable | perl -e '
my %values, %help, @uniq;

$maxlen = 0;

while (<>) {
  if (/^mca:/) {
    chop;
    @tmp = split(":");
    $length = length($tmp[4]);
    $maxlen = ($length, $maxlen)[$length < $maxlen];
    $values{$tmp[4]} = 1;
    if ($tmp[5] eq "help") {
      $help{$tmp[4]} = $tmp[6];
      $help{$tmp[4]} =~ s/:/\\:/g;
    }
  }
}

for $exclude (split(" ", "'"$excl"'")) {
  delete $values{$exclude};
}

for $key (sort(keys %values)) {
  print $key;
  if ($help{$key} ne "") {
    print " " x ($maxlen - length($key)) . "- $help{$key}";
  }

  print "\n";
}' 2> /dev/null
}

# given a framework (type) name print a list of components
_get_mca_component_names() {
    # components are printed by default (-a is not needed)
    ompi_info -l 9 --parsable | grep "mca:$1:" | perl -e '
my %values;

while (<>) {
  if (/^mca:/) {
    @tmp = split(":");
    $values{$tmp[2]} = 1;
  }
}

@uniq = keys %values;
print join(" ", sort(@uniq));'
}

_get_mpirun_switches() {
    mpirun --help 2>&1 | sed 's/^\s*//g' | egrep '^-' | cut -d' ' -f1 | tr '|\n' ' '
}

# get enumerator values for a variable
_get_enum_values() {
    ompi_info -a -l 9 --parsable | grep ":$1:" | perl -e '
my @values;

while (<>) {
  chop;
  @tmp = split(":");

  if ($tmp[5] eq "enumerator") {
    $values[++$#values] = $tmp[7];
    $values[++$#values] = $tmp[8];
  }
}
print join(" ", @values);'
}

# remove items from $1 that are also in $2. lists must be sorted
_set_remove () {
    comm -23 <(echo $1 | sort | tr " " "\n") <(echo $2 | sort | tr " " "\n") 2>/dev/null
}

# find mca parameters specified on the command line (prevent duplicates)
_find_mca_parameters() {
    for ((i = 1; i < COMP_CWORD; i++)) ; do
        # the subcommand is the first item on the command line that isn't module or a switch
        if test ${COMP_WORDS[$i]##*-} == "mca" -a $i -lt $((COMP_CWORD-1)) ; then
            echo ${COMP_WORDS[$i+1]}
        fi
    done
}

# check for matches that have the search term somewhere in the name (but not the description)
_fuzzy_search() {
    for match in $2 ; do
        if [[ ${match%%[[:space:]]*} =~ .*"$1".* ]] ; then
            echo $match
        fi
    done
}

# mpirun/orterun completion
_mpirun() {
    local cur prv tb switches already_specified all_variables avail_variables enumerations save_IFS
    local disable_description=yes

    # Enable descriptions if we can detect the completion type (so we can strip the description off
    # if there is only one result. This is the case for normal completion (ASCII 9) or successive
    # completion (?: ASCII 63)
    if test "$COMP_TYPE" = "9" -o "$COMP_TYPE" = "63" ; then
        disable_description=no
    fi

    COMPREPLY=()
    if test $COMP_CWORD -gt 1 ; then
        tb=${COMP_WORDS[COMP_CWORD-2]}
    else
        tb=""
    fi

    if test $COMP_CWORD -gt 0 ; then
        prv=${COMP_WORDS[COMP_CWORD-1]}
    else
        prv=""
    fi

    cur=${COMP_WORDS[COMP_CWORD]}

    if test "${prv##*-}" = "mca" ; then
        # Complete variable name

        # Temporarily change IFS to newline
        save_IFS=$IFS
        IFS="
"

        # Remove mca parameters already on the command line
        avail_variables=($(_get_mca_variable_names "$(_find_mca_parameters)"))

	IFS=" "
        # Check if we should disable descriptions
        if test "$disable_description" = "yes" ; then
            avail_variables=(${avail_variables[*]%%[[:space:]]*})
        fi
	IFS="
"

        # Return a fuzzy-search of the mca parameter names
        COMPREPLY=($(_fuzzy_search "$cur" "${avail_variables[*]}"))

        # Remove the description if this is the last item
        if test ${#COMPREPLY[@]} -eq 1 ; then
            completion=${COMPREPLY[0]}
            COMPREPLY=("${COMPREPLY[0]%%[[:space:]]*}")
        fi
        IFS=$save_IFS
    elif test "${tb##*-}" = "mca" ; then
        # Complete variable value

        # Check if the variable is a selection variable (no _ in the name)
        if test "${prv#_}" = "${prv}" ; then
            # component selection variable, find available components (removing ones already selected)
            enumerations=($(_set_remove "$(_get_mca_component_names $prv)" "$(echo $cur | tr ',' '\n')"))

            # Prepend the current selection if there is one
            if test "${cur%,*}" = "${cur}" ; then
                COMPREPLY=($(_fuzzy_search "${cur##*,}" "${enumerations[*]}"))
            else
                COMPREPLY=($(_fuzzy_search "${cur##*,}" "${enumerations[*]}" | sed "s/^/${cur%,*},/g"))
            fi
        else
            enumerations=($(_get_enum_values "$prv"))
            COMPREPLY=($(_fuzzy_search "$cur" "${enumerations[*]}"))
        fi
    elif test "${prv}" = "--bind-to" ; then
        COMPREPLY=($(compgen -W "none hwthread core socket numa board" -- "$cur"))
    elif test "${prv}" = "--map-by" -o "${prv}" = "-map-by" ; then
        COMPREPLY=($(compgen -W "slot hwthread core socket numa board node" -- "$cur"))
    elif test "${prv##*-}" = "hostfile" ; then
        COMPREPLY=($(compgen -f -- "$cur"))
    elif test "${cur:0:1}" = "-" ; then
        switches=$(_get_mpirun_switches)
        COMPREPLY=($(compgen -W "$switches" -- "$cur"))
    else
        COMPREPLY=($(compgen -f -- "$cur"))
    fi

    return 0
}

complete -o nospace -F _mpirun mpirun
complete -o nospace -F _mpirun orterun
complete -o nospace -F _mpirun mpiexec
